core/tests/performance_framework/perf_domain.py

```diff
@@ -0,0 +1,156 @@
+                            the main document
+            domLoading: the Unix time the DOM started loading
+            fetchStart: the Unix time the page began loading
+            loadEventEnd: the Unix time the page finished loading
+            requestStart: the Unix time the request started
+            responseStart: the Unix time the response started
+            responseEnd: the Unix time the request finished
+    """
+
+    TIMING_PROPERTIES = [
+        'loadEventEnd', 'fetchStart', 'domComplete', 'domInteractive',
+        'responseEnd', 'requestStart'
+    ]
+
+    def __init__(self, page_session_stats=None, page_session_timings=None):
+        self.page_session_stats = page_session_stats
+        self.page_session_timings = page_session_timings
+
+        self.page_load_timings = {}
+        if self.page_session_timings:
+            self.page_load_timings = page_session_timings['timing']
+
+        self._validate()
+
+    def _validate(self):
+        """Validates various properties of a PageSessionMetrics object."""
+        if not self.page_session_stats and not self.page_session_timings:
+            raise utils.ValidationError(
+                'Expected one of page_session_stats or page_session_timings '
+                'to be provided.')
+
+        if self.page_session_stats:
+            if 'log' not in self.page_session_stats:
+                raise utils.ValidationError(
+                    'Expected the page load stats to have a \'log\' entry')
+
+            if 'entries' not in self.page_session_stats['log']:
+                raise utils.ValidationError(
+                    'Expected the log entry of the page load stats to include '
+                    'an additional \'entries\' element')
+
+            for entry in self.page_session_stats['log']['entries']:
+                if '_error' in entry['response']:
+                    raise utils.ValidationError(
+                        'Expected a valid server response, found server '
+                        'not reachable.')
+
+            if self.get_request_count() == 0:
+                raise utils.ValidationError(
+                    'Expected the log entry of the page load stats to include '
+                    'a positive number of entries.')
+
+            if self.get_total_page_size_bytes() == 0:
+                raise utils.ValidationError(
+                    'Expected the total size of a page including all its '
+                    'resources to be positive.')
+
+        if self.page_session_timings:
+            for timing_prop in self.TIMING_PROPERTIES:
+                if timing_prop not in self.page_load_timings:
+                    raise utils.ValidationError(
+                        'Expected the timing entry of the page load timings to '
+                        'include %s property' % timing_prop)
+
+                if timing_prop in self.page_session_timings:
+                    if len(str(self.page_session_timings[timing_prop])) < 13:
```

Comment:
I know the specs specify that it should be in milliseconds. But, I added it just to be sure.